src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.cpp

```diff
@@ -16,33 +23,70 @@
 /// \cond
 namespace NewtonianEuler {
 namespace Sources {
+template <>
+void VortexPerturbation<2>::apply() const noexcept {}
 
-void VortexPerturbation::apply(
-    const gsl::not_null<Scalar<DataVector>*> mass_density_source,
-    const gsl::not_null<tnsr::I<DataVector, 3>*> momentum_density_source,
-    const gsl::not_null<Scalar<DataVector>*> energy_density_source,
-    const Scalar<DataVector>& vortex_mass_density,
-    const tnsr::I<DataVector, 3>& vortex_momentum_density,
-    const Scalar<DataVector>& vortex_energy_density,
-    const Scalar<DataVector>& vortex_pressure,
-    const Scalar<DataVector>& vortex_velocity_z,
-    const Scalar<DataVector>& dz_vortex_velocity_z) noexcept {
-  get(*mass_density_source) =
-      get(vortex_mass_density) * get(dz_vortex_velocity_z);
+template <>
+void VortexPerturbation<3>::apply(
+    const gsl::not_null<Scalar<DataVector>*> source_mass_density_cons,
+    const gsl::not_null<tnsr::I<DataVector, 3>*> source_momentum_density,
+    const gsl::not_null<Scalar<DataVector>*> source_energy_density,
+    const NewtonianEuler::Solutions::IsentropicVortex<3>& vortex,
+    const tnsr::I<DataVector, 3>& x, const double time) const noexcept {
+  const size_t number_of_grid_points = get<0>(x).size();
+  const auto vortex_primitives = vortex.variables(
+      x, time,
+      tmpl::list<Tags::MassDensity<DataVector>, Tags::Velocity<DataVector, 3>,
+                 Tags::SpecificInternalEnergy<DataVector>,
+                 Tags::Pressure<DataVector>>{});
+  Variables<
+      tmpl::list<::Tags::TempScalar<0>, ::Tags::TempI<1, 3, Frame::Inertial>,
+                 ::Tags::TempScalar<2>>>
+      temp_buffer(number_of_grid_points);
+  auto& vortex_mass_density_cons = get<::Tags::TempScalar<0>>(temp_buffer);
+  auto& vortex_momentum_density =
+      get<::Tags::TempI<1, 3, Frame::Inertial>>(temp_buffer);
+  auto& vortex_energy_density = get<::Tags::TempScalar<2>>(temp_buffer);
+
+  NewtonianEuler::ConservativeFromPrimitive<3>::apply(
+      make_not_null(&vortex_mass_density_cons),
+      make_not_null(&vortex_momentum_density),
+      make_not_null(&vortex_energy_density),
+      get<Tags::MassDensity<DataVector>>(vortex_primitives),
+      get<Tags::Velocity<DataVector, 3, Frame::Inertial>>(vortex_primitives),
+      get<Tags::SpecificInternalEnergy<DataVector>>(vortex_primitives));
+
+  // We save the precomputed value of dv_z/dz in source_mass_density_cons
+  // in order to save an allocation
+  get(*source_mass_density_cons) =
+      vortex.perturbation_amplitude() *
+      vortex.deriv_of_perturbation_profile(get<2>(x));
```

Comment:
A simpler way to do this would have been to reuse the allocation returned by `deriv_of_perturbation_profile`.  (Assign it to a variable and then `*=` the amplitude.)  (Just a comment.  Not requesting a change.)